WcaOnRails/lib/results_validators/events_rounds_validator.rb

```diff
@@ -0,0 +1,121 @@
+          [c.id, c]
+        end
+      ]
+
+      results_by_competition_id.each do |competition_id, results_for_comp|
+        competition = competitions[competition_id]
+
+        check_main_event(competition, results_for_comp)
+
+        check_events_match(competition, results_for_comp)
+
+        if competition.has_rounds?
+          check_rounds_match(competition, results_for_comp)
+        end
+      end
+      self
+    end
+
+    private
+
+    def check_main_event(competition, results)
+      unless results.map(&:eventId).uniq.include?("333")
+        @warnings << ValidationWarning.new(:events, competition.id,
+                                           CHOOSE_MAIN_EVENT_WARNING)
+      end
+    end
+
+    def check_events_match(competition, results)
+      # Check for missing/unexpected events
+      # As events must be validated by WCAT, any missing or unexpected event should lead to an error.
+      expected = competition.events.map(&:id)
+      real = results.map(&:eventId).uniq
+
+      (real - expected).each do |event_id|
+        @errors << ValidationError.new(:events, competition.id,
+                                       UNEXPECTED_RESULTS_ERROR,
+                                       event_id: event_id)
+      end
+
+      (expected - real).each do |event_id|
+        @warnings << ValidationWarning.new(:events, competition.id,
+                                           MISSING_RESULTS_WARNING,
+                                           event_id: event_id)
+      end
+    end
+
+    def check_rounds_match(competition, results)
+      # Check that rounds match what was declared.
+      # This function automatically casts combined rounds to regular rounds if everyone has met the cutoff.
+
+      expected_rounds_by_ids = Hash[competition.competition_events.map(&:rounds).flatten.map { |r| ["#{r.event.id}-#{r.round_type_id}", r] }]
+
+      expected = expected_rounds_by_ids.keys
+      real = results.map { |r| "#{r.eventId}-#{r.roundTypeId}" }.uniq
+      unexpected = real - expected
+      missing = expected - real
+
+      missing.each do |round_id|
+        event_id, round_type_id = round_id.split("-")
+        equivalent_round_id = "#{event_id}-#{RoundType.toggle_cutoff(round_type_id)}"
+        if unexpected.include?(equivalent_round_id)
+          unexpected.delete(equivalent_round_id)
+          round = expected_rounds_by_ids[round_id]
+          if round.round_type.combined?
+            # NOTE: we cannot know if everyone legitimately cleared the cutoff,
+            # or if the cutoff was removed during the competition and not
+            # updated on the website's schedule. So we just consider it fine,
+            # but we have to update the expected round information so that we have
+            # a valid round_info when checking individual results later.
+            expected_rounds_by_ids[equivalent_round_id] = expected_rounds_by_ids.delete(round_id)
```

Comment:
This seems to not actually do what it should? Looks like the previous instance variable was used by the individual results validation code, which was separated earlier and without taking this into account, as far as I can understand.

Reply:
Hum yes part of this was moved over to the GenericValidator!